### PR TITLE
Framework: Make C++17 the default

### PR DESCRIPTION
**Pull request by sebrowne**
Also remove support for C++14.

User Support Ticket(s) or Story Referenced:
https://github.com/trilinos/Trilinos/issues/11356

@trilinos/framework

## Motivation
C++17 is required for the upcoming Kokkos integration.

## Related Issues
#11356

### CODE REVIEW

**Comment by sebrowne** [comment]
I didn't search exhaustively, but I think this does what @ndellingwood is asking for in #11356 

@ccober6 This is what came up earlier today.  It will break any customers who can't use C++17 effectively immediately (meaning they won't be able to integrate newer versions of Trilinos).

**Review by ndellingwood — APPROVED** [review]
Thanks @sebrowne ! Changes look good to me

**Comment by sebrowne** [comment]
@rppawlo I know you asked about the timeline surrounding this yesterday

**Comment by ndellingwood** [comment]
Just a note that without @sebrowne's change here that current builds with Stokhos enabled and an unspecified CMAKE_CXX_STANDARD (resulting in c++14 being enabled) or if explicitly requesting c++14 will fail to compile. A hack that may work around this is something like:

```c++
diff --git a/packages/stokhos/src/sacado/kokkos/vector/Sacado_MP_Vector.hpp b/packages/stokhos/src/sacado/kokkos/vector/Sacado_MP_Vector.hpp
index ac6543de688..9a55c9c2eea 100644
--- a/packages/stokhos/src/sacado/kokkos/vector/Sacado_MP_Vector.hpp
+++ b/packages/stokhos/src/sacado/kokkos/vector/Sacado_MP_Vector.hpp
@@ -269,6 +269,7 @@ namespace Sacado {
        */
       KOKKOS_INLINE_FUNCTION
       Vector(std::initializer_list<value_type> l) : s(l.size(), l.begin()) {
+#ifndef KOKKOS_ENABLE_CXX14
         if constexpr (Storage::is_static) {
           const auto         lsz = static_cast<ordinal_type>(l.size());
           const ordinal_type  sz = this->size();
@@ -282,6 +283,7 @@ namespace Sacado {
             }
           }
         }
+#endif
       }
 
       //! Destructor
```

**Comment by rppawlo** [comment]
@sebrowne - you might also want to add 23 to the allowed standards while you are changing this file. It is supported in cmake since 3.20.

**Comment by sebrowne** [comment]
Planning on merging this ~ 01-09-2023

**Review by ccober6 — APPROVED** [review]

**Comment by alanw0** [comment]
@sebrowne we put a stk snapshot into develop yesterday which uses c++17 and it broke a bunch of trilinos builds. We were under the impression that c++17 was already legal based on emails which we might have mis-intrepreted?
I guess we should revert that snapshot.
When is c++17 expected to be legal in trilinos?

**Comment by cgcgcg** [comment]
@alanw0 I don't think you need to revert if merging this PR fixes the issues.

**Comment by sebrowne** [comment]
@alanw0 which builds failed?  Once this PR is merged, C++17 is mandatory (and automatic), so said builds will either break for other reasons that aren't your fault, or they will start passing with your changes because they're currently still set to C++14.  Either way, I don't think your changes are an issue.

**Comment by cgcgcg** [comment]
https://testing.sandia.gov/cdash/index.php?project=Trilinos for the dashboard with the failing builds.

**Comment by sebrowne** [comment]
Okay that's kind of what I thought.  No worries @alanw0, we can wait and see what happens once this PR is merged.

**Comment by alanw0** [comment]
> Okay that's kind of what I thought. No worries @alanw0, we can wait and see what happens once this PR is merged.

To be clear, are you suggesting I wait on reverting the stk commit?

**Comment by sebrowne** [comment]
Correct, I do NOT think you need to revert the commit

**Comment by jhux2** [comment]
Since #11461 assumes c++17, shouldn't it have failed PR testing?

**Comment by sebrowne** [comment]
@jhux2 No, because PR testing is explicitly enabling C++17 (this happened during the compiler upgrade in early November).  Since then, we've been in a state where C++17 is fully supported via PR testing, but was not actually required by TriBITS (that is, TriBITS would allow for C++17 to not be enabled, but C++17 features may have been present in the code, leading to exactly what happened here).  We were delaying requiring C++17 until we resolved @glhenni's reported issue in #11356, as well as by request of @rppawlo for EMPIRE to get further along towards C++17 support.

TL;DR: The possibility of this exact situation was known by myself and @ccober6 when we did the autotester change in November, and we accepted that possibility until we merged this PR.  Merging this PR should prevent this issue from now on by enabling C++17 by default.

**Comment by jhux2** [comment]
@sebrowne Got it, thank you for the explanation!
